### DISCUSSION TITLE
How to find number of subscriptions on jetstream

### DISCUSSION THREAD

**Discussion by YashasAnand** [nats-io/nats-server#6117] [Q&A] [answered]
I am using jetsream but instead of consuming messages I am subscribing to subjects with queue groups attached.

If i do nats consumer list i do not get list of consumers. If so how do i get number of subscribers?

I have tried to look in management console on 8222 port, but i do not get details like if its my own subscriptios or some server subscriptions as the numbers are not matching.

**Comment by Jarema** [comment]
In JetStream, you cannot subscribe to Stream Subjects without a Consumer, so you definately should see consumers in consumer list.

In CLI, you can use `nats stream report` and `nats consumer report` to get good info on JetStream.
In monitoring APIs, you can [add flags to list more details](https://docs.nats.io/running-a-nats-service/nats_admin/monitoring#jetstream-information), especially `streams` and `consumers`.

**Comment by YashasAnand** [comment]
Thank you for your reply @Jarema , pls see the attached screenshots of nats consumer report & nats stream report

I am using below snippet of code to do these operations. Is this a bug?

Pls refer to code below:
```csharp
NatsOpts natsOptions = new NatsOpts
{
    Url = connectionInfo.BrokerUrl, 
    Name = connectionInfo.ClientName,
};

_natsConnection = new NatsConnection(natsOptions);

_natsJSContext = new NatsJSContext(_natsClient);

await foreach (var msg in _natsClient.SubscribeAsync(natsConsumerOptions.TopicName,natsConsumerOptions.ConsumerGroup,serializer:natsMesgSerializer))
{
    _logger.Log($"Received message on Subscriber {natsConsumerOptions.TopicName} in group {natsConsumerOptions.ConsumerGroup}: {msg.Data}");
    _ = Task.Run(async () => await ProcessMessageAsyncNew(natsConsumerOptions,consumerOptions,onMessageReceived,msg));
}

```


<img width="665" alt="Stream_report" src="https://github.com/user-attachments/assets/703062e3-63ee-4ea3-b650-9ecbd3582f3e">
<img width="599" alt="consumer_report" src="https://github.com/user-attachments/assets/837c658e-1eef-4fe3-ac4a-44c3194606be">

**Reply by Jarema** [reply]
I see you're craeting `_natsJSContext`, but you are not actually using it. I guess it's a bug in your code, but I'm not C# engineer.
@mtmk can help here for sure.

You can also use Pull Consumers, like [here](https://nats-by-example-g1str9e3k-connecteverything.vercel.app/examples/jetstream/pull-consumer/csharp), to achieve the same goal:

**Comment by YashasAnand** [comment]
Ah I see, @Jarema thank you for pointing this out, but then how are the messages still persisiting to stream if i am not using jetstream context?

**Reply by Jarema** [reply, marked as answer]
you can publish messages to the stream with just Core NATS, but you will not get ACK, not a backpressure control (so it's not a good idea).
JetStream Publish sends a request instead, that will expect ack back.

**Comment by YashasAnand** [comment]
@Jarema thank you so much for your help. Much Aprreciated !!

**Reply by YashasAnand** [reply]
@Jarema , in a jetstream consumer with limit based retention, if i want 6 consumers for a subject , i am currently creating 6 consumers with same name & same filter subject ( which is subject name) . In this scenario, jetstream is considering this as a single entity but internally i can see that messages are being distributed to all 6 consumers.

So my question is, am i actually creating 6 consumers or only 1?

**Reply by Jarema** [reply]
You are creating one consumer.
In case of pull consumers, it can load balance the messages across clients that are bound to that consumer.

If you had 6 consumers, each would get their copy of the message on the stream.

The reason you end up with 1 - is that name has to be unique.

**Reply by YashasAnand** [reply]
@Jarema Thank you for your reply, what i wanted to confirm is if internally jetstream knows that there are 6 pull consumers bound to subject, & by increasing the number of consumers I can actually see more messages getting consumed even though it shows only 1 consumer in jetstream.

Or do i get some additional throughput if i actually create 6 consumers with different names & subject filters by adding some extra logic to filtering?

**Reply by Jarema** [reply]
JetStream does not internally "know" anything - It's client side action to create a consumer and then use it.
If you're using old JS API , it might be hidden behind `js.Subscribe`. What client are you using?

If you will create 6 seperate consumers, each will get their own copy of messages in the stream.
You can treat consumers as a "view" into the stream. They track the progress through the stream and acks done by clients.
They are server side constructs.
https://docs.nats.io/nats-concepts/jetstream/consumers

**Reply by YashasAnand** [reply]
@Jarema , I am using c# v2 client and below is my code. Lets say for subject orders.burgers with consumer name as burgers & i need 6 consumers for this subject.  I run this logic in a loop & create 6 consumers with same name & same filter subject (which is topicName) in limit based retention stream.

But i can see only 1 consumer spotted in jetstream. I just want to confirm if i am actually creating 6 consumers or not. or do i need to change my logic

```csharp
private readonly NatsConnection _natsClient;
private readonly NatsJSContext _natsJSContext;
NatsOpts natsOptions = new NatsOpts
{
    Url = connectionInfo.BrokerUrl, 
    Name = connectionInfo.ClientName,
};

_natsConnection = new NatsConnection(natsOptions);
_natsJSContext = new NatsJSContext(_natsClient);
var ConsumerConfig = new ConsumerConfig
{
    Name = natsConsumerOptions.ConsumerName,
    DurableName = natsConsumerOptions.ConsumerName,
    AckPolicy = ConsumerConfigAckPolicy.Explicit,
    DeliverPolicy = ConsumerConfigDeliverPolicy.Last,
    FilterSubject = natsConsumerOptions.TopicName,
    AckWait = TimeSpan.FromMinutes(1)
};
var consumer = await _natsJSContext!.CreateOrUpdateConsumerAsync(natsConsumerOptions.StreamName, ConsumerConfig);
_logger.Log($"Started Consumer {natsConsumerOptions.ConsumerName} with FilterSubject {natsConsumerOptions.TopicName}");
await foreach (var jsMsg in consumer.ConsumeAsync(serializer:natsMesgSerializer))
{
    //Processing logic
}
```

**Reply by Jarema** [reply]
1. Consumers are identified by a name. You cannot have multiple consumers with the same name. You have 1 consumer with 6 clients binding to it. Which is ok.
2. Each consumer tracks its progress through the stream. If you would have multiple consumers, you would get each stream sequence for each consumer.

**Reply by YashasAnand** [reply]
"You have 1 consumer with 6 clients binding to it. " . How does this work exactly? & which is preferred, to have 6 consumers or 1 consumers with 6 bindings for higher consumption rate?

**Reply by Jarema** [reply]
Please read the provided docs link and my previous responses.

If you bind 6 clients to 1 consumer, server will round-robin messages across them.
Having 6 consumers would result in different outcome - sending every messages to each, instead of round robin.

**Reply by YashasAnand** [reply]
@Jarema thank you for the response, what im asking is if i have 6 consumers with different names & different subject filters such that i make sure for a subject (orders.burgers becomes orders.burgers.1 with consumer 1 listening only to orders.burgers.1) one message goes only to one consumer using different subject filters, Will i get a better performance in this or will i get a better performance in round-robin binding consumer.

**Reply by Jarema** [reply]
That depends a lot on subject cardinality, replication factor, etc.
The best approach is to bench your specific scenario. NATS CLI has great tooling capabilities with `nats bench`.

**Reply by YashasAnand** [reply]
Thank you very much for your valuable inputs & time @Jarema
